src/config.rs

```diff
@@ -266,14 +276,30 @@ fn default_convert_dos_to_unix() -> bool {
 }
 
 impl Config {
-    pub fn from_file(path: impl AsRef<Path>) -> Result<Self> {
-        let contents = fs::read_to_string(path.as_ref()).context(ReadConfigFileSnafu {
-            filename: path.as_ref(),
+    pub fn from_path(path: &PathBuf) -> Result<Self> {
+        let filename = if path.is_dir() {
+            path.join("mujmap.toml")
+        } else {
+            path.clone()
+        };
+
+        let contents = fs::read_to_string(&filename).context(ReadConfigFileSnafu {
+            filename: filename.clone(),
         })?;
-        let config: Self = toml::from_str(contents.as_str()).context(ParseConfigFileSnafu {
-            filename: path.as_ref(),
+        let mut config: Self = toml::from_str(contents.as_str()).context(ParseConfigFileSnafu {
+            filename: filename.clone(),
```

Comment:
`filename.as_ref()` should work here I think. Snafu works in mysterious ways (and by that I mean it creates a bunch of objects that accept `AsRef<X>` where X is an owned object that is eventually cloned into the error type proper in the case of the error occurring)

Reply:
Plain `as_ref()` seems to be ambiguous:

```
error[E0283]: type annotations needed
   --> src/config.rs:287:32
    |
287 |             filename: filename.as_ref(),
    |                       ---------^^^^^^--
    |                       |        |
    |                       |        cannot infer type for type parameter `T` declared on the trait `AsRef`
    |                       this method call resolves to `&T`
    |
    = note: multiple `impl`s satisfying `PathBuf: AsRef<_>` found in the `std` crate:
            - impl AsRef<Path> for PathBuf;
            - impl AsRef<std::ffi::OsStr> for PathBuf;
help: use the fully qualified path for the potential candidates
    |
287 |             filename: <PathBuf as AsRef<Path>>::as_ref(filename),
    |                       +++++++++++++++++++++++++++++++++        ~
287 |             filename: <PathBuf as AsRef<std::ffi::OsStr>>::as_ref(filename),
    |                       ++++++++++++++++++++++++++++++++++++++++++++        ~
```

This diff gets it going:

```diff
diff --git src/config.rs src/config.rs
index 343a946..12aa9a9 100644
--- src/config.rs
+++ src/config.rs
@@ -2,7 +2,7 @@ use serde::Deserialize;
 use snafu::prelude::*;
 use std::{
     fs, io,
-    path::PathBuf,
+    path::{Path, PathBuf},
     process::{Command, ExitStatus},
     string::FromUtf8Error,
 };
@@ -284,10 +284,10 @@ impl Config {
         };
 
         let contents = fs::read_to_string(&filename).context(ReadConfigFileSnafu {
-            filename: filename.clone(),
+            filename: <PathBuf as AsRef<Path>>::as_ref(&filename),
         })?;
         let mut config: Self = toml::from_str(contents.as_str()).context(ParseConfigFileSnafu {
-            filename: filename.clone(),
+            filename: <PathBuf as AsRef<Path>>::as_ref(&filename),
         })?;
 
         // In directory mode, if paths aren't offered then we use the config dir itself.
```

If it were me I'd leave it, because the overheads of an additional `clone()` in an error path seem insignificant next to the readability hit. But, your house, your rules. Do you have a strong preference?

Reply:
Of course as soon as I posted I saw the ref/deref hoop jumping, and replaced it all with: `filename: &filename`. Never mind.